### PR TITLE
gh-70795: Rework RLock documentation

### PR DESCRIPTION
**Pull request by samatjain**
Attempted to simultaneously reduce verbosity, while more descriptively describing behavior.

Fix links (RLock acquire/release previously linking to Lock acquire/release, seems like bad copy pasta).

* Issue: gh-70795

### CODE REVIEW

**Comment by samatjain** [comment]
cc @CAM-Gerlach thanks for draft review!

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
To lock the lock, a thread calls the lock's :meth:`~RLock.acquire` method.  To unlock
```

Avoid any ambiguity about whether "it" is the thread or the lock.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
Threads call a lock's :meth:`~RLock.acquire` method to lock it,
and its :meth:`~Lock.release` method to unlock it. Reentrant locks
```

Alternatively, this would fix the above issue and also reduce duplication and redundancy.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
support the :ref:`context management protocol <with-locks>` so :keyword:`with` can be used.
```

Fix syntax mistake (triple instead of double backticks); format and link `with`.

Comment (CAM-Gerlach) [reply Doc/library/threading.rst]: Also, `with` can be used how? I'm guessing maybe

```suggestion
support the :ref:`context management protocol <with-locks>` so :keyword:`with` can be used
to acquire and release the lock automatically in a block of code.
```

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
:meth:`~RLock.acquire`/:meth:`~RLock.release` must be used in pairs: each acquire
```

Minor clarification

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
call release as many times the lock has been acquired can lead to deadlock.
```

Fix typo and apply Greg's suggestions

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
            lock, or *timeout*, if set to a positive float value.
```

Be more explicit here and incorporate the additional information below.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]: This should be a bulleted rather than ordered list, as the latter implies a sequence of steps with an explict order rather than multiple different options/paths with no fixed, predetermined order.

Additionally, I would suggest swapping the first and second item, since the current case "2" is the simplest.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst:637]:
```suggestion
            return immediately. This is the difference between :class:`Lock` and
            :class:`!RLock`; :class:`Lock` handles this case the same as the previous,
            blocking until the lock can be acquired.
            Internally, the "recursion level" is atomically
```

Be much more explicit and clear as to how these two differ.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
      When invoked with the *blocking* argument set to ``False``:
```

Just jump directly to what it actually means, like the previous ``True`` case

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
      not acquired.
```

Un-acquired would be interpreted as verb here, where this is in fact a noun (a state)

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
         3. If the same thread owns the lock, acquire the lock again and return
```

Fix typo

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]: Same here—this should be a bulleted list, with the simplest case (2.) first.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
```

Remove this redundant sentence; incorporated above.

File: Doc/library/threading.rst
Comment (gpshead) [inline Doc/library/threading.rst]: wording, perhaps:

... call release as many times as the lock has been acquired can lead to deadlock.

File: Doc/library/threading.rst
Comment (gpshead) [inline Doc/library/threading.rst]: Lets not mention the internal implementation details.  get rid of the "Internally, ..." sentence.

File: Doc/library/threading.rst
Comment (gpshead) [inline Doc/library/threading.rst]: comma instead of .

File: Doc/library/threading.rst
Comment (gpshead) [inline Doc/library/threading.rst]: ... (i.e. if not blocking or the timeout was reached) ...

File: Doc/library/threading.rst
Comment (gpshead) [inline Doc/library/threading.rst]: "if this method is called on a non-acquired lock." perhaps?

**Review by gpshead — DISMISSED** [review]
some rewording suggestions.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
      In all cases, if the thread was able to acquire the lock, return ``True``.
```

As above

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
      *timeout* is reached or is non-positive) return ``False``.
```

Fix grammar, add formatting and clarify case of timeout being non-positive

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
      may lead to deadlock. Consider using :class:`!RLock` as a context manager rather than
```

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]: I suggest making this a `.. seealso::` block after the summary line rather than buried at the end.

**Review by CAM-Gerlach — COMMENTED** [review]
Thanks! Various suggested changes and fixed.

**Standard reminder: You can directly apply all the suggestions you want in one go by going to [`Files changed`](https://github.com/python/cpython/pull/103853/files) -> Clicking <kbd>Add to batch</kbd> on each suggestion -> When done, clicking <kbd>Commit</kbd>**

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
            return immediately. This is the difference between :class:`Lock` and
            :class:`!RLock`; :class:`Lock` handles this case the same as the previous,
            blocking until the lock can be acquired.
```

Be much more explicit and clear as to how these two differ. Also, remove the last sentence, which is an internal implementation detail.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [reply Doc/library/threading.rst]: Yeah, noticed that as well. Implemented in my suggestions

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [reply Doc/library/threading.rst]: Yup; I added this in my suggestion

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [reply Doc/library/threading.rst]: Yup, fixed in my suggestion

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [reply Doc/library/threading.rst]: I changed it to just "if this method is called when the lock is not acquired" but we can revise it further if needed.

File: Doc/library/threading.rst
Comment (samatjain) [reply Doc/library/threading.rst]: Removed.

Note: we talk about other internal details in the beginning of the RLock, as well as the doc for `release`, that this PR otherwise did not touch.

File: Doc/library/threading.rst
Comment (samatjain) [reply Doc/library/threading.rst]: Fixed.

File: Doc/library/threading.rst
Comment (samatjain) [reply Doc/library/threading.rst]: Fixed.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
and its :meth:`~Lock.release` method to unlock it.

.. note::

  Reentrant locks support the :ref:`context management protocol <with-locks>`,
  so it is recommended to use :keyword:`with` instead of manually calling
  :meth:`~RLock.acquire` and :meth:`~RLock.release`
  to handle acquiring and releasing the lock for a block of code.
```

Call this out in a note

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
   :ref:`Using RLock as a context manager <with-locks>` is recommended
   over manual :meth:`!acquire` and :meth:`release` calls whenever practical.
```

Here's how I would phrase this, and also move it down past the first summary sentence in `RLock.acquire()` ("Acquire a lock, blocking or non-blocking")

**Review by CAM-Gerlach — COMMENTED** [review]
A couple followup comments

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [inline Doc/library/threading.rst]:
```suggestion
         :ref:`Using RLock as a context manager <with-locks>`
            Recommended over manual :meth:`!acquire` and :meth:`release` calls
            whenever practical.
```

**Review by CAM-Gerlach — APPROVED** [review]
LGTM from my side; thanks @samatjain ! I'll leave it to @gpshead to re-review the content.

File: Doc/library/threading.rst
Comment (CAM-Gerlach) [reply Doc/library/threading.rst]: Seems like this is resolved, so I'll mark this as resolved and merge this, and we can iterate further if necessary in a followup. Thanks!

**Pull request merged**
